File: GGAndroid_2.0/GGCore/GGUtils/src/main/java/io/ganguo/utils/util/Bitmaps.java
package io.ganguo.utils.util;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.media.ExifInterface;
import android.os.Build;
import android.util.Base64;
import android.widget.ImageView;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

import io.ganguo.utils.bean.Globals;
import io.ganguo.utils.util.log.Logger;

/**
 * bitmap处理工具类
 * <p>
 * Created by HulkYao on 4/11/15.
 */
public class Bitmaps {

    private Bitmaps() {
        throw new Error(Globals.ERROR_MSG_UTILS_CONSTRUCTOR);
    }

    /**
     * 保存为文件
     * (完成后回收bitmap)
     *
     * @param bitmap
     * @param to
     * @param format
     * @param quality
     */
    public static void toFile(Bitmap bitmap, File to, Bitmap.CompressFormat format, int quality) {
        Validates.notNull(bitmap);
        try {
            FileOutputStream fileOutputStream = new FileOutputStream(to);
            bitmap.compress(format, quality, fileOutputStream);
            fileOutputStream.flush();
            fileOutputStream.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            Logger.e("handleImageToFile failed:", e);
        } catch (IOException e) {
            e.printStackTrace();
            Logger.e("handleImageToFile failed:", e);
        }
    }

    /**
     * 保存为PNG
     * (完成后回收bitmap)
     *
     * @param bitmap
     * @param to
     * @param quality
     */
    public static void toPNGFile(Bitmap bitmap, File to, int quality) {
        toFile(bitmap, to, Bitmap.CompressFormat.PNG, quality);
    }

    /**
     * 保存为JPEG
     * (完成后回收bitmap)
     *
     * @param bitmap
     * @param to
     * @param quality
     */
    public static void toJPEGFile(Bitmap bitmap, File to, int quality) {
        toFile(bitmap, to, Bitmap.CompressFormat.JPEG, quality);
    }

    /**
     * 获取bitmap占用的内存体积(byte)
     *
     * @param bitmap byte
     * @return
     */
    public static long size(Bitmap bitmap) {
        Validates.notNull(bitmap);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            return bitmap.getAllocationByteCount();
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR1) {
            return bitmap.getByteCount();
        } else {
            return (long) bitmap.getRowBytes() * bitmap.getHeight();
        }
    }

    /**
     * 旋转bitmap
     *
     * @param bitmap
     * @param degrees
     * @return bitmap
     */
    public static Bitmap rotate(Bitmap bitmap, int degrees) {
        Matrix matrix = new Matrix();
        matrix.postRotate(degrees);
        // bitmap new
        return Bitmap.createBitmap(bitmap, 0, 0,
                bitmap.getWidth(), bitmap.getHeight(), matrix, true);
    }

    /**
     * 旋转图片至正确位置
     *
     * @param bitmap
     * @param path
     * @return
     * @throws IOException
     */
    public static Bitmap rotateToNormal(Bitmap bitmap, String path) {
        int angle = readDegree(path);
        if (angle == 0) {
            return bitmap;
        }
        return rotate(bitmap, angle);

    }

    /**
     * 用于压缩file类型文件大小
     *
     * @param inputBitmap
     * @param output
     * @param outPutSize
     * @return
     */
    public static File compress(Bitmap inputBitmap, File output, int outPutSize) {
        int quality = 100;
        // 最小quality是10
        while (quality > 10) {
            quality -= 5;
            Bitmaps.toJPEGFile(inputBitmap, output, quality);

            long fileSize = output.length() / 1024;
            if (fileSize <= outPutSize) {
                break;
            }
            Logger.i("scaleBitmap" + "quality," + quality +
                    " output.length() / 1024 :" + (output.length() / 1024) +
                    " outPutSize:" + outPutSize);
        }
        return output;
    }

    /**
     * 等比缩放
     *
     * @param from
     * @return
     */
    public static Bitmap uniformScale(File from) {
        BitmapFactory.Options newOpts = new BitmapFactory.Options();
        // 开始读入图片，此时把options.inJustDecodeBounds 设回true了
        newOpts.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(from.getPath(), newOpts);

        newOpts.inJustDecodeBounds = false;
        int w = newOpts.outWidth;
        int h = newOpts.outHeight;

        int be = computeSize(w, h);
        while (true) {
            try {
                // 设置缩放比例
                newOpts.inSampleSize = be;
                return BitmapFactory.decodeFile(from.getPath(), newOpts);
            } catch (OutOfMemoryError outOfMemoryError) {
                Logger.e("java.lang.OutOfMemoryError,current be:" + be);
                be++;
            }
        }
    }


    /**
     * 分辨率的压缩算法
     * <p>
     * 接近微信图片保存时图片分辨率算法
     * https://github.com/Curzibn/Luban/blob/master/library/src/main/java/top/zibin/luban/Engine.java
     *
     * @param srcWidth
     * @param srcHeight
     * @return
     */
    private static int computeSize(int srcWidth, int srcHeight) {
        int mSampleSize;

        srcWidth = srcWidth % 2 == 1 ? srcWidth + 1 : srcWidth;
        srcHeight = srcHeight % 2 == 1 ? srcHeight + 1 : srcHeight;

        srcWidth = srcWidth > srcHeight ? srcHeight : srcWidth;
        srcHeight = srcWidth > srcHeight ? srcWidth : srcHeight;

        double scale = ((double) srcWidth / srcHeight);

        if (scale <= 1 && scale > 0.5625) {
            if (srcHeight < 1664) {
                mSampleSize = 1;
            } else if (srcHeight >= 1664 && srcHeight < 4990) {
                mSampleSize = 2;
            } else if (srcHeight >= 4990 && srcHeight < 10240) {
                mSampleSize = 4;
            } else {
                mSampleSize = srcHeight / 1280 == 0 ? 1 : srcHeight / 1280;
            }
        } else if (scale <= 0.5625 && scale > 0.5) {
            mSampleSize = srcHeight / 1280 == 0 ? 1 : srcHeight / 1280;
        } else {
            mSampleSize = (int) Math.ceil(srcHeight / (1280.0 / scale));
        }

        return mSampleSize;
    }

    /**
     * 拿图片路径缩略图
     *
     * @param path 图片的路径
     * @return Bitmap
     */
    public static Bitmap getThumbPic(String path) {
        ExifInterface exif = null;
        try {
            exif = new ExifInterface(path);
        } catch (IOException e) {
            e.printStackTrace();
            Logger.e("failed to get ExifInterface", e);
        }
        if (exif == null || exif.getThumbnail() == null) {
            return null;
        }
        byte[] imageData = exif.getThumbnail();
        return BitmapFactory.decodeByteArray(imageData, 0, imageData.length);
    }

    /**
     * 读取图片属性：旋转的角度
     *
     * @param path 图片绝对路径
     * @return degree旋转的角度
     */
    public static int readDegree(String path) {
        int degree = 0;
        try {
            ExifInterface exifInterface = new ExifInterface(path);
            int orientation = exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL);
            switch (orientation) {
                case ExifInterface.ORIENTATION_NORMAL:
                    degree = 0;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_90:
                    degree = 90;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_180:
                    degree = 180;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_270:
                    degree = 270;
                    break;
            }
        } catch (IOException e) {
            Logger.e("failed to readPictureDegree", e);
        }
        return degree;
    }

    /**
     * 毛玻璃效果
     *
     * @param from
     * @param radius
     * @param canReuseInBitmap
     * @return bitmap
     */
    public static Bitmap doBlur(Bitmap from, int radius, boolean canReuseInBitmap) {
        Validates.notNull(from);

        Bitmap bitmap;
        if (canReuseInBitmap) {
            bitmap = from;
        } else {
            bitmap = from.copy(from.getConfig(), true);
        }

        if (radius < 1) {
            return (null);
        }

        int w = bitmap.getWidth();
        int h = bitmap.getHeight();

        int[] pix = new int[w * h];
        bitmap.getPixels(pix, 0, w, 0, 0, w, h);

        int wm = w - 1;
        int hm = h - 1;
        int wh = w * h;
        int div = radius + radius + 1;

        int r[] = new int[wh];
        int g[] = new int[wh];
        int b[] = new int[wh];
        int rsum, gsum, bsum, x, y, i, p, yp, yi, yw;
        int vmin[] = new int[Math.max(w, h)];

        int divsum = (div + 1) >> 1;
        divsum *= divsum;
        int dv[] = new int[256 * divsum];
        for (i = 0; i < 256 * divsum; i++) {
            dv[i] = (i / divsum);
        }

        yw = yi = 0;

        int[][] stack = new int[div][3];
        int stackpointer;
        int stackstart;
        int[] sir;
        int rbs;
        int r1 = radius + 1;
        int routsum, goutsum, boutsum;
        int rinsum, ginsum, binsum;

        for (y = 0; y < h; y++) {
            rinsum = ginsum = binsum = routsum = goutsum = boutsum = rsum = gsum = bsum = 0;
            for (i = -radius; i <= radius; i++) {
                p = pix[yi + Math.min(wm, Math.max(i, 0))];
                sir = stack[i + radius];
                sir[0] = (p & 0xff0000) >> 16;
                sir[1] = (p & 0x00ff00) >> 8;
                sir[2] = (p & 0x0000ff);
                rbs = r1 - Math.abs(i);
                rsum += sir[0] * rbs;
                gsum += sir[1] * rbs;
                bsum += sir[2] * rbs;
                if (i > 0) {
                    rinsum += sir[0];
                    ginsum += sir[1];
                    binsum += sir[2];
                } else {
                    routsum += sir[0];
                    goutsum += sir[1];
                    boutsum += sir[2];
                }
            }
            stackpointer = radius;

            for (x = 0; x < w; x++) {

                r[yi] = dv[rsum];
                g[yi] = dv[gsum];
                b[yi] = dv[bsum];

                rsum -= routsum;
                gsum -= goutsum;
                bsum -= boutsum;

                stackstart = stackpointer - radius + div;
                sir = stack[stackstart % div];

                routsum -= sir[0];
                goutsum -= sir[1];
                boutsum -= sir[2];

                if (y == 0) {
                    vmin[x] = Math.min(x + radius + 1, wm);
                }
                p = pix[yw + vmin[x]];

                sir[0] = (p & 0xff0000) >> 16;
                sir[1] = (p & 0x00ff00) >> 8;
                sir[2] = (p & 0x0000ff);

                rinsum += sir[0];
                ginsum += sir[1];
                binsum += sir[2];

                rsum += rinsum;
                gsum += ginsum;
                bsum += binsum;

                stackpointer = (stackpointer + 1) % div;
                sir = stack[(stackpointer) % div];

                routsum += sir[0];
                goutsum += sir[1];
                boutsum += sir[2];

                rinsum -= sir[0];
                ginsum -= sir[1];
                binsum -= sir[2];

                yi++;
            }
            yw += w;
        }
        for (x = 0; x < w; x++) {
            rinsum = ginsum = binsum = routsum = goutsum = boutsum = rsum = gsum = bsum = 0;
            yp = -radius * w;
            for (i = -radius; i <= radius; i++) {
                yi = Math.max(0, yp) + x;

                sir = stack[i + radius];

                sir[0] = r[yi];
                sir[1] = g[yi];
                sir[2] = b[yi];

                rbs = r1 - Math.abs(i);

                rsum += r[yi] * rbs;
                gsum += g[yi] * rbs;
                bsum += b[yi] * rbs;

                if (i > 0) {
                    rinsum += sir[0];
                    ginsum += sir[1];
                    binsum += sir[2];
                } else {
                    routsum += sir[0];
                    goutsum += sir[1];
                    boutsum += sir[2];
                }

                if (i < hm) {
                    yp += w;
                }
            }
            yi = x;
            stackpointer = radius;
            for (y = 0; y < h; y++) {
                // Preserve alpha channel: ( 0xff000000 & pix[yi] )
                pix[yi] = (0xff000000 & pix[yi]) | (dv[rsum] << 16)
                        | (dv[gsum] << 8) | dv[bsum];

                rsum -= routsum;
                gsum -= goutsum;
                bsum -= boutsum;

                stackstart = stackpointer - radius + div;
                sir = stack[stackstart % div];

                routsum -= sir[0];
                goutsum -= sir[1];
                boutsum -= sir[2];

                if (x == 0) {
                    vmin[y] = Math.min(y + r1, hm) * w;
                }
                p = x + vmin[y];

                sir[0] = r[p];
                sir[1] = g[p];
                sir[2] = b[p];

                rinsum += sir[0];
                ginsum += sir[1];
                binsum += sir[2];

                rsum += rinsum;
                gsum += ginsum;
                bsum += binsum;

                stackpointer = (stackpointer + 1) % div;
                sir = stack[stackpointer];

                routsum += sir[0];
                goutsum += sir[1];
                boutsum += sir[2];

                rinsum -= sir[0];
                ginsum -= sir[1];
                binsum -= sir[2];

                yi += w;
            }
        }

        bitmap.setPixels(pix, 0, w, 0, 0, w, h);

        return (bitmap);
    }

    /**
     * recycle bitmap
     * <p/>
     *
     * @param bitmap
     */
    public static void recycle(Bitmap bitmap) {
        if (bitmap == null) {
            return;
        }
        if (!bitmap.isRecycled()) {
            bitmap.recycle();
        }
    }

    /**
     * recycle drawable
     * -> recycle(Bitmap bitmap)
     *
     * @param drawable
     */
    public static void recycle(Drawable drawable) {
        if (drawable == null) {
            return;
        }
        if (drawable instanceof BitmapDrawable) {
            recycle(((BitmapDrawable) drawable).getBitmap());
        }
    }

    /**
     * recycle imageView
     * -> recycle(Bitmap bitmap)
     *
     * @param imageView
     */
    public static void recycle(ImageView imageView) {
        if (imageView == null) {
            return;
        }
        Drawable drawable = imageView.getDrawable();
        imageView.setImageBitmap(null);
        imageView.setImageDrawable(null);
        recycle(drawable);
    }

    /**
     * bitmap转为base64
     *
     * @param bitmap
     * @return
     */
    public static String toBase64(Bitmap bitmap) {
        String result = null;
        ByteArrayOutputStream baos = null;
        try {
            if (bitmap != null) {
                baos = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);

                baos.flush();
                baos.close();

                byte[] bitmapBytes = baos.toByteArray();
                result = Base64.encodeToString(bitmapBytes, Base64.DEFAULT);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (baos != null) {
                    baos.flush();
                    baos.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return result;
    }
}

File: GGAndroid_3.0/core/support/utils/src/main/java/io/ganguo/utils/util/Bitmaps.kt
package io.ganguo.utils.util

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.media.ExifInterface
import android.os.Build
import android.widget.ImageView
import io.ganguo.log.core.Logger
import io.ganguo.utils.bean.Globals
import java.io.*
import java.net.URL
import kotlin.math.ceil

/**
 * bitmap处理工具类
 *
 *
 * Created by HulkYao on 4/11/15.
 */
class Bitmaps private constructor() {
    init {
        throw Error(Globals.ERROR_MSG_UTILS_CONSTRUCTOR)
    }

    companion object {

        /**
         * 保存为文件
         * (完成后回收bitmap)
         *
         * @param bitmap
         * @param to
         * @param format
         * @param quality
         */
        @JvmStatic
        fun toFile(bitmap: Bitmap, to: File, format: Bitmap.CompressFormat, quality: Int) {
            Validates.notNull(bitmap)
            try {
                val fileOutputStream = FileOutputStream(to)
                bitmap.compress(format, quality, fileOutputStream)
                fileOutputStream.flush()
                fileOutputStream.close()
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                Logger.e("handleImageToFile failed:", e.message)
            } catch (e: IOException) {
                e.printStackTrace()
                Logger.e("handleImageToFile failed:", e.message)
            }

        }

        /**
         * 保存为PNG
         * (完成后回收bitmap)
         *
         * @param bitmap
         * @param to
         * @param quality
         */
        @JvmStatic
        fun toPNGFile(bitmap: Bitmap, to: File, quality: Int) {
            toFile(bitmap, to, Bitmap.CompressFormat.PNG, quality)
        }

        /**
         * 保存为JPEG
         * (完成后回收bitmap)
         *
         * @param bitmap
         * @param to
         * @param quality
         */
        @JvmStatic
        fun toJPEGFile(bitmap: Bitmap, to: File, quality: Int) {
            toFile(bitmap, to, Bitmap.CompressFormat.JPEG, quality)
        }

        /**
         * 获取bitmap占用的内存体积(byte)
         *
         * @param bitmap byte
         * @return
         */
        @JvmStatic
        fun size(bitmap: Bitmap): Long {
            Validates.notNull(bitmap)
            return when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT -> bitmap.allocationByteCount.toLong()
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR1 -> bitmap.byteCount.toLong()
                else -> bitmap.rowBytes.toLong() * bitmap.height
            }
        }

        /**
         * 旋转bitmap
         *
         * @param bitmap
         * @param degrees
         * @return bitmap
         */
        @JvmStatic
        fun rotate(bitmap: Bitmap, degrees: Int): Bitmap {
            val matrix = Matrix()
            matrix.postRotate(degrees.toFloat())
            // bitmap new
            return Bitmap.createBitmap(bitmap, 0, 0,
                    bitmap.width, bitmap.height, matrix, true)
        }

        /**
         * 旋转图片至正确位置
         *
         * @param bitmap
         * @param path
         * @return
         * @throws IOException
         */
        @JvmStatic
        fun rotateToNormal(bitmap: Bitmap, path: String): Bitmap {
            val angle = readDegree(path)
            return if (angle == 0) {
                bitmap
            } else rotate(bitmap, angle)

        }

        /**
         * 用于压缩file类型文件大小
         *
         * @param inputBitmap
         * @param output
         * @param outPutSize
         * @return
         */
        @JvmStatic
        fun compress(inputBitmap: Bitmap, output: File, outPutSize: Int): File {
            var quality = 100
            // 最小quality是10
            while (quality > 10) {
                quality -= 5
                toJPEGFile(inputBitmap, output, quality)
                val fileSize = output.length() / 1024
                if (fileSize <= outPutSize) {
                    break
                }
            }
            return output
        }

        /**
         * 等比缩放
         *
         * @param from
         * @return
         */
        @JvmStatic
        fun uniformScale(from: File): Bitmap {
            val newOpts = BitmapFactory.Options()
            // 开始读入图片，此时把options.inJustDecodeBounds 设回true了
            newOpts.inJustDecodeBounds = true
            BitmapFactory.decodeFile(from.path, newOpts)

            newOpts.inJustDecodeBounds = false
            val w = newOpts.outWidth
            val h = newOpts.outHeight

            var be = computeSize(w, h)
            while (true) {
                try {
                    // 设置缩放比例
                    newOpts.inSampleSize = be
                    return BitmapFactory.decodeFile(from.path, newOpts)
                } catch (outOfMemoryError: OutOfMemoryError) {
                    Logger.e("java.lang.OutOfMemoryError,current be:$be")
                    be++
                }

            }
        }


        /**
         * 分辨率的压缩算法
         *
         *
         * 接近微信图片保存时图片分辨率算法
         * https://github.com/Curzibn/Luban/blob/master/library/src/main/java/top/zibin/luban/Engine.java
         *
         * @param srcWidth
         * @param srcHeight
         * @return
         */
        private fun computeSize(srcWidth: Int, srcHeight: Int): Int {
            var width = srcWidth
            var height = srcHeight
            val mSampleSize: Int
            width = if (width % 2 == 1) width + 1 else width
            height = if (height % 2 == 1) height + 1 else height
            width = if (width > height) height else width
            height = if (width > height) width else height
            val scale = width.toDouble() / height
            if (scale <= 1 && scale > 0.5625) {
                if (height < 1664) {
                    mSampleSize = 1
                } else if (height in 1664..4989) {
                    mSampleSize = 2
                } else if (height in 4990..10239) {
                    mSampleSize = 4
                } else {
                    mSampleSize = if (height / 1280 == 0) 1 else height / 1280
                }
            } else if (scale <= 0.5625 && scale > 0.5) {
                mSampleSize = if (height / 1280 == 0) 1 else height / 1280
            } else {
                mSampleSize = ceil(height / (1280.0 / scale)).toInt()
            }
            return mSampleSize
        }

        /**
         * 拿图片路径缩略图
         *
         * @param path 图片的路径
         * @return Bitmap
         */
        fun getThumbBitmap(path: String): Bitmap? {
            var exif: ExifInterface? = null
            try {
                exif = ExifInterface(path)
            } catch (e: IOException) {
                e.printStackTrace()
                Logger.e("failed to get ExifInterface", e.message)
            }

            if (exif == null || exif.thumbnail == null) {
                return null
            }
            val imageData = exif.thumbnail
            return BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
        }

        /**
         * 读取图片属性：旋转的角度
         *
         * @param path 图片绝对路径
         * @return degree旋转的角度
         */
        fun readDegree(path: String): Int {
            var degree = 0
            try {
                val exifInterface = ExifInterface(path)
                val orientation = exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
                when (orientation) {
                    ExifInterface.ORIENTATION_NORMAL -> degree = 0
                    ExifInterface.ORIENTATION_ROTATE_90 -> degree = 90
                    ExifInterface.ORIENTATION_ROTATE_180 -> degree = 180
                    ExifInterface.ORIENTATION_ROTATE_270 -> degree = 270
                }
            } catch (e: IOException) {
                Logger.e("failed to readPictureDegree", e.message)
            }

            return degree
        }


        /**
         * 转换资源编码
         *
         * @param bitmap      位图
         * @param needRecycle 是否回收
         * @param quality     图片质量，上限为100
         * @return
         */
        fun bitmapToByteArray(bitmap: Bitmap, quality: Int = 100, needRecycle: Boolean = true): ByteArray {
            val output = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, quality, output)
            if (needRecycle) {
                bitmap.recycle()
            }
            val result = output.toByteArray()
            try {
                output.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return result
        }

        /**
         * @param bitmap     原图
         * @param edgeLength 希望得到的正方形部分的边长
         * @return 缩放截取正中部分后的位图。
         */
        fun centerSquareScaleBitmap(bitmap: Bitmap?, edgeLength: Int): Bitmap? {
            if (null == bitmap || edgeLength <= 0) {
                return null
            }
            var result: Bitmap = bitmap
            val widthOrg = bitmap.width
            val heightOrg = bitmap.height
            if (widthOrg > edgeLength && heightOrg > edgeLength) {
                //压缩到一个最小长度是edgeLength的bitmap
                val longerEdge = edgeLength * Math.max(widthOrg, heightOrg) / Math.min(widthOrg, heightOrg)
                val scaledWidth = if (widthOrg > heightOrg) longerEdge else edgeLength
                val scaledHeight = if (widthOrg > heightOrg) edgeLength else longerEdge
                val scaledBitmap: Bitmap

                try {
                    scaledBitmap = Bitmap.createScaledBitmap(bitmap, scaledWidth, scaledHeight, true)
                } catch (e: Exception) {
                    return null
                }

                //从图中截取正中间的正方形部分。
                val xTopLeft = (scaledWidth - edgeLength) / 2
                val yTopLeft = (scaledHeight - edgeLength) / 2

                try {
                    result = Bitmap.createBitmap(scaledBitmap, xTopLeft, yTopLeft, edgeLength, edgeLength)
                    scaledBitmap.recycle()
                } catch (e: Exception) {
                    return null
                }

            }
            return result
        }

        /**
         * 处理大图导致oom的问题，返回bitmap
         *
         * @param imagePath
         * @return
         */
        fun dealImageSize(imagePath: String): Bitmap? {
            val options = BitmapFactory.Options()
            options.inJustDecodeBounds = true
            BitmapFactory.decodeFile(imagePath, options)
            val height = options.outHeight * 720 / options.outWidth
            // 计算图片缩放比例
            options.inSampleSize = computeSampleSize(options, -1, 720 * height)

            options.outWidth = 720
            options.outHeight = height
            // 这样才能真正的返回一个Bitmap给你
            options.inJustDecodeBounds = false
            return BitmapFactory.decodeFile(imagePath, options)
        }


        /**
         * 计算图片大小，Android提供了一种动态计算BitmapFactory.Options() inSampleSize的方法。
         *
         * @param bitmap 位图
         * @return int
         */
        fun computeSampleSize(options: BitmapFactory.Options,
                              minSideLength: Int, maxNumOfPixels: Int): Int {
            val initialSize = computeInitialSampleSize(options, minSideLength,
                    maxNumOfPixels)

            var roundedSize: Int
            if (initialSize <= 8) {
                roundedSize = 1
                while (roundedSize < initialSize) {
                    roundedSize = roundedSize shl 1
                }
            } else {
                roundedSize = (initialSize + 7) / 8 * 8
            }

            return roundedSize
        }

        /**
         * 计算图片大小
         *
         * @param bitmap 位图
         * @return int
         */
        private fun computeInitialSampleSize(options: BitmapFactory.Options,
                                             minSideLength: Int, maxNumOfPixels: Int): Int {
            val w = options.outWidth.toDouble()
            val h = options.outHeight.toDouble()

            val lowerBound = if (maxNumOfPixels == -1)
                1
            else
                Math.ceil(Math.sqrt(w * h / maxNumOfPixels)).toInt()
            val upperBound = if (minSideLength == -1)
                128
            else
                Math.min(Math.floor(w / minSideLength),
                        Math.floor(h / minSideLength)).toInt()

            if (upperBound < lowerBound) {
                // return the larger one when there is no overlapping zone.
                return lowerBound
            }

            return if (maxNumOfPixels == -1 && minSideLength == -1) {
                1
            } else if (minSideLength == -1) {
                lowerBound
            } else {
                upperBound
            }
        }

        /**
         * 计算图片大小
         *
         * @param bitmap 位图
         * @return int
         */
        @JvmStatic
        fun getBitmapSize(bitmap: Bitmap): Int {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {     //API 19
                return bitmap.allocationByteCount
            }
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR1) {//API 12
                bitmap.byteCount
            } else bitmap.rowBytes * bitmap.height
        }

        /**
         * 网络资源图片转bitmap
         *
         * @param imageUrl 网络资源图片链接
         * @return Bitmap
         */
        @JvmStatic
        fun httpImageUrlToBitmap(imageUrl: String): Bitmap? {
            val bitmap: Bitmap
            val inPut: InputStream
            val outPut: BufferedOutputStream
            try {
                inPut = BufferedInputStream(URL(imageUrl).openStream(), 1024)
                val dataStream = ByteArrayOutputStream()
                outPut = BufferedOutputStream(dataStream, 1024)
                val b = ByteArray(1024)
                var read: Int = inPut.read(b)
                while (read != -1) {
                    outPut.write(b, 0, read)
                }
                outPut.flush()
                val data = dataStream.toByteArray()
                bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
                return bitmap
            } catch (e: IOException) {
                e.printStackTrace()
                return null
            }

        }


        /**
         * 图片文件路径转Bitmap位图
         *
         * @param path       图片路径
         * @param edgeLength 希望得到的正方形部分的边长
         * @return Bitmap
         */
        @JvmStatic
        fun filePathToBitmap(path: String, edgeLength: Int): Bitmap? {
            return fileToBitmap(File(path), edgeLength)
        }


        /**
         * 图片文件转Bitmap位图
         *
         * @param file       文件
         * @param edgeLength 希望得到的正方形部分的边长
         * @return Bitmap
         */
        @JvmStatic
        fun fileToBitmap(file: File, edgeLength: Int): Bitmap? {
            val thumbBmp: Bitmap?
            if (!file.exists()) {
                return null
            }
            //因为一些图片太大，容易导致OOM，所以先处理下
            val bitmap = dealImageSize(file.absolutePath) ?: return null
            //设置缩略图
            thumbBmp = centerSquareScaleBitmap(bitmap, edgeLength)
            if (!bitmap.isRecycled) {
                // 回收图片所占的内存
                bitmap.recycle()
                // 提醒系统及时回收
                System.gc()
            }
            return thumbBmp
        }


        /**
         * recycle bitmap
         *
         * @param bitmap
         */
        @JvmStatic
        fun recycle(bitmap: Bitmap?) {
            if (bitmap == null) {
                return
            }
            if (!bitmap.isRecycled) {
                bitmap.recycle()
            }
        }

        /**
         * recycle drawable
         * -> recycle(Bitmap bitmap)
         *
         * @param drawable
         */
        @JvmStatic
        fun recycle(drawable: Drawable?) {
            if (drawable == null) {
                return
            }
            if (drawable is BitmapDrawable) {
                recycle(drawable.bitmap)
            }
        }

        /**
         * recycle imageView
         * -> recycle(Bitmap bitmap)
         *
         * @param imageView
         */
        @JvmStatic
        fun recycle(imageView: ImageView?) {
            if (imageView == null) {
                return
            }
            val drawable = imageView.drawable
            imageView.setImageBitmap(null)
            imageView.setImageDrawable(null)
            recycle(drawable)
        }
    }

}
